README.md
<div id="top"></div>
<p align="center">
    <a href="https://www.npmjs.com/package/reactaccelerometervalue">
        <img src="https://img.shields.io/npm/v/reactaccelerometervalue"
            alt="NPM Package"></a>
    <a href="https://github.com/maheshrao98/React-Accelerometer-Value/blob/main/LICENSE.txt">
          <img src="https://img.shields.io/github/license/maheshrao98/React-Accelerometer-Value"
              alt="License"/></a>
    <a href="https://twitter.com/raomahesh98">
        <img src="https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Ftwitter.com%2Fraomahesh98"
            alt="follow on Twitter"></a>
</p>
<div align="center">
  <h3 align="center">React-Accelerometer-Value</h3>

  <p align="center">
    The most easiest way to obtain the accelerometer and orientation values of your device in React JS.
    <br />
    <br />
    <a href="https://reactaccelerometertest.netlify.app">View Demo</a>
    ·
    <a href="https://github.com/maheshrao98/React-Accelerometer-Value/issues">Report Bug</a>
    ·
    <a href="https://github.com/maheshrao98/React-Accelerometer-Value/issues">Request Feature</a>
  </p>
</div>

This is an improvised version of [React Accelerometer](https://www.npmjs.com/package/react-accelerometer) and almost similar to React Native [Expo Sensor](https://www.npmjs.com/package/expo-sensors) . It will run efficiently in the latest versions of React. I have also made notable changes in the way the accelerometer value is calculated according to the latest [Sensor API Documentation](https://developer.mozilla.org/en-US/docs/Web/API/Sensor_APIs). 

## Installation
```js
npm install --save reactaccelerometervalue
// or
yarn add reactaccelerometervalue
```

## Usage
The React-Accelerometer-Value is a parent component with the value of orientation and accelerometer values are passed as children values. Therefore, to obtain the values the best way is to wrap your component that you need the value to display in as a Child component inside the ReactAccelerometerValue.

### In App.js
```js
import React from "react";
import ReactAccelorometerValue from 'reactaccelerometervalue'
import List from './List'

class App extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return (
      <div >
        <ReactAccelorometerValue>
          <List></List>
        </ReactAccelorometerValue>
      </div>
    );
  }
}
export default App
```
### In List.js
```js
import React,{Component} from 'react';

class List extends Component {
    constructor(props) {
        super(props)
    }
    render(){
        console.log(this.props)
        return(
            <div>
                <ul>
                <li>Is it landscape:{this.props.state.landscape.toString()} </li>
                <li>x:{this.props.state.x}</li>
                <li>y:{this.props.state.y}</li>
                <li>z:{this.props.state.z}</li>
                <li>rotation alpha:{this.props.state.rotation.alpha}</li>
                <li>rotation beta:{this.props.state.rotation.beta}</li>
                <li>rotation gamma:{this.props.state.rotation.gamma}</li>
                </ul>
            </div>
        )
    }
}

export default List;
```
<p align="right">(<a href="#top">Back to top</a>)</p>

## Test
```js
npm test
// or
yarnpkg test
```

## Roadmap
- [x] Improvise to support latest React versions
- [x] Resolve IOS 13+ Gyroscope Permission Request Issues
- [ ] Build an example app

<p align="right">(<a href="#top">Back to top</a>)</p>

## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#top">Back to top</a>)</p>

## License
Distributed under the MIT License. See [LICENSE.txt](https://github.com/maheshrao98/React-Accelerometer-Value/blob/main/LICENSE.txt) for more information.

<p align="right">(<a href="#top">Back to top</a>)</p>

## Contact

Author - [@raomahesh98](https://twitter.com/raomahesh98) - raomahesh610@gmail.com

Project Link: [https://github.com/maheshrao98/React-Accelerometer-Value](https://github.com/maheshrao98/React-Accelerometer-Value)

<p align="right">(<a href="#top">Back to top</a>)</p>


HISTORY.md
Legend:
- [B]: Breaking
- [F]: Fix
- [I]: Improvement

### 0.1.8 (October 7th 2021)
- [F] Resolve IOS 13+ Gyroscope Permission Request Issues

### 0.1.7 (October 5th 2021)
- [I] Initial release.

src/lib/index.js
import React,{Component, cloneElement, Children } from 'react'

class ReactAccelorometerValue extends Component {
  constructor(props) {
    super(props);
    this.state = {
      x: 0,
      y: 0,
      z: 0,
      rotation: {
        alpha: 0,
        beta: 0,
        gamma: 0
      },
      landscape: false
    };
  };

  componentDidMount(){
    this.handleOrientation();
    if (typeof DeviceMotionEvent.requestPermission === 'function') {
      DeviceMotionEvent.requestPermission()
        .then(permissionState => {
          if (permissionState === 'granted') {
            window.addEventListener('devicemotion', this.handleOrientation);
          }
        })
        .catch(console.error);
    } else {
      // handle regular non iOS 13+ devices
      window.addEventListener('devicemotion', this.handleAcceleration);
    }
    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
      DeviceOrientationEvent.requestPermission()
        .then(permissionState => {
          if (permissionState === 'granted') {
            window.addEventListener('deviceorientation', this.handleOrientation);
          }
        })
        .catch(console.error);
    } else {
      // handle regular non iOS 13+ devices
      window.addEventListener('orientationchange', this.handleOrientation);
    }
  }

  componentWillUnmount(){
    window.removeEventListener('devicemotion', this.handleAcceleration);
    window.removeEventListener('orientationchange', this.handleOrientation);
  }

  handleOrientation = () => {
    var orientation = window.orientation;
    this.setState({ landscape: orientation === 90 || orientation === -90 });
  }

  handleAcceleration = (event) => {
    var landscape = this.state.landscape;
    var acceleration = event.accelerationIncludingGravity;
    var rotation = event.rotationRate || 0;
    var x = acceleration.x;
    var y = acceleration.y;
    var z = acceleration.z;
    this.setState({
      rotation: rotation,
      x: landscape ? y : x,
      y: landscape ? x : y,
      z: z,
      landscape: this.state.landscape
    });
  }

  render() {
    return (
        <div>
        {
            Children.map(this.props.children, child => {
                return cloneElement(child, { state: this.state });
            })
        }
        </div>
    );
  }
}

export default ReactAccelorometerValue;

src/lib/ReactAccelorometerValue.js
import React,{Component, cloneElement, Children } from 'react'

class ReactAccelorometerValue extends Component {
  constructor(props) {
    super(props);
    this.state = {
      x: null,
      y: null,
      z: null,
      rotation: {
        alpha: null,
        beta: null,
        gamma: null
      },
      landscape: false
    };
  };

  componentDidMount(){
    this.handleOrientation();
    if (typeof DeviceMotionEvent.requestPermission === 'function') {
      DeviceMotionEvent.requestPermission()
        .then(permissionState => {
          if (permissionState === 'granted') {
            window.addEventListener('devicemotion', this.handleOrientation);
          }
        })
        .catch(console.error);
    } else {
      // handle regular non iOS 13+ devices
      window.addEventListener('devicemotion', this.handleAcceleration);
    }
    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
      DeviceOrientationEvent.requestPermission()
        .then(permissionState => {
          if (permissionState === 'granted') {
            window.addEventListener('deviceorientation', this.handleOrientation);
          }
        })
        .catch(console.error);
    } else {
      // handle regular non iOS 13+ devices
      window.addEventListener('orientationchange', this.handleOrientation);
    }
  }

  componentWillUnmount(){
    window.removeEventListener('devicemotion', this.handleAcceleration);
    window.removeEventListener('orientationchange', this.handleOrientation);
  }

  handleOrientation = () => {
    var orientation = window.orientation;
    this.setState({ landscape: orientation === 90 || orientation === -90 });
  }

  handleAcceleration = (event) => {
    var landscape = this.state.landscape;
    var acceleration = event.accelerationIncludingGravity;
    var rotation = event.rotationRate || null;
    var x = acceleration.x;
    var y = acceleration.y;
    var z = acceleration.z;
    
    this.setState({
      rotation: rotation,
      x: landscape ? y : x,
      y: landscape ? x : y,
      z: z,
      landscape: this.state.landscape.toString()
    });
  }

  render() {
    return (
        <div>
        {
            Children.map(this.props.children, child => {
                return cloneElement(child, { state: this.state });
            })
        }
        </div>
    );
  }
}

export default ReactAccelorometerValue;

src/App.js
import React,{Component, cloneElement, Children } from 'react'
import './App.css';

class ReactAccelorometerValue extends Component {
  constructor(props) {
    super(props);
    this.state = {
      x: 0,
      y: 0,
      z: 0,
      rotation: {
        alpha: 0,
        beta: 0,
        gamma: 0
      },
      landscape: false,
      noaccess: false
    };
  };

  componentDidMount(){
    this.handleOrientation();
    if (typeof DeviceMotionEvent.requestPermission === 'function' || typeof DeviceOrientationEvent.requestPermission === 'function') {
      this.setState({noaccess:true})
      window.addEventListener('orientationchange', this.handleOrientation);
    } else {
      // handle regular non iOS 13+ devices
      window.addEventListener('devicemotion', this.handleAcceleration);
      window.addEventListener('orientationchange', this.handleOrientation);
    }
  }

  handleSubmit = () => {
    console.log("hello")
    window.DeviceMotionEvent.requestPermission()
      .then(permissionState => {
        if (permissionState === 'granted') {
          window.addEventListener('devicemotion', this.handleOrientation);
          this.setState({noaccess:false})
        }
      })
      .catch(console.error);
  }

  componentWillUnmount(){
    window.removeEventListener('devicemotion', this.handleAcceleration);
    window.removeEventListener('orientationchange', this.handleOrientation);
  }

  handleOrientation = () => {
    var orientation = window.orientation;
    this.setState({ landscape: orientation === 90 || orientation === -90 });
  }

  handleAcceleration = (event) => {
    var landscape = this.state.landscape;
    var acceleration = event.accelerationIncludingGravity;
    var rotation = event.rotationRate || 0;
    var x = acceleration.x;
    var y = acceleration.y;
    var z = acceleration.z;
    this.setState({
      rotation: rotation,
      x: landscape ? y : x,
      y: landscape ? x : y,
      z: z,
      landscape: this.state.landscape
    });
  }

  render() {
    const renderBanner = () => {
      if(this.state.noaccess) {
        return (
          <div className="DeviceMotionRequestBanner">
            <form onSubmit={this.handleSubmit}>
              <h3>Enable DeviceMotion on your device?</h3>
              <p>This app requires you to enable the DeviceMotion event on yout device.</p>
              <div style={{position:"relative",zIndex:"1"}}>
                <button className="DeviceMotionRequestButton">Enable</button>
              </div>
            </form>
          </div>
        )
      }
    }

    return (
        <div>
        {renderBanner()}
        {
            Children.map(this.props.children, child => {
                return cloneElement(child, { state: this.state });
            })
        }
        </div>
    );
  }
}

export default ReactAccelorometerValue;